Document debugging with Solr Admin Query Screen (#51)
metabrainz/mb-solr: HACKING.md

The Debugger's Guide to MusicBrainz Solr
========================================

> This file gives tips to debug MusicBrainz (advanced or not) indexed search based on Solr.

Reviewer: ```suggestion
This file gives tips to debug MusicBrainz indexed search (basic and advanced) using Solr.
```

Table of contents
-----------------

<!-- toc -->

- [Setup](#setup)
- [Solr logs](#solr-logs)
- [Solr Admin](#solr-admin)
  * [Query screen](#query-screen)

<!-- tocstop -->

Setup
-----

Set a development setup with [`musicbrainz-docker`](https://github.com/metabrainz/musicbrainz-docker).

> It allows to run any released or development version of

Reviewer: ```suggestion
This can be used to run any released or development version of
```

Author: Applied as c124a0c

in Apache Solr Reference Guide 7.7 for more information.

### Query screen

The query screen allows to reproduce and tweak search queries made through MBS.

To follow the same example as in the above section “[Solr logs](#solr-logs)”,
switch the Core Selector to “`release`” and click on “Query” to get to
<http://localhost:8983/solr/#/release/query>.

Then start filling the form as follows:
* Set the first field “Request-Handler (qt)” (qt)
  to the value for `path` in the log record, with a leading slash:
  `/advanced`
* Set the second field “q” (query string)
  to the text you typed for the search
  (that is the unescaped value for`q` in the log record):
  `(Downward Spiral) AND format:CD`
* Select `xml` in the field `wt` (writer type) for the most readable presentation.

From here, different combinations can be choosed depending on your needs:

* Check `debugQuery` to have a detailed score computation (using search fields and boosts) for each result.

* Set `fl` (field list) to
  either `mbid,score` for the least detailed list of results
  or `*,score` for the most detailed list of results

> * Restrict to a single result by setting “Raw Query Parameters” to:

Reviewer: replace with:

If you expect another item to show up in the search results but it doesn't, you can ask solr to return the score logging for that item and compare it with the scores of the search results.

Author: Applied in commit 458125c

  `explainOther=mbid:fe78827f-e402-42a0-ab63-a3fd11c12cc4`

> Alternatively to get the same output format as in MB Search API
> (only since it is incompatible with showing any details),
> Unselect any value in the field `wt` (writer type) and
> set “Raw Query parameters” to either `wt=mbjson` or `wt=mbxml`.

Reviewer: ```suggestion
Alternatively to get the same output format as in MB Search API,m
Unselect any value in the field `wt` (writer type) and
set “Raw Query parameters” to either `wt=mbjson` or `wt=mbxml`.
These formats are incompatible with the `fl`, `debugQuery`, and `explainOther` flags, and setting
these options will not change the output.
```

Author: Applied as b630fcc and amended with f966475

> 

Reviewer: new section:

Some parameters are automatically set on all queries based on the core configuration. 

Some parameters are set based on the request handler: <https://github.com/metabrainz/mbsssss/blob/master/common/requestHandler-advanced.xml>. These are identified as "invariants", which means that the parameters will always be set to this value even if you specify a different value in the admin.

Some parameters for ranking are set with defaults and can be overridden: <https://github.com/metabrainz/mbsssss/blob/master/recording/conf/request-params.xml>
These can be changed by copying them into the relevant fields in the admin query panel to see the resulting change in ranking order.

Author: Added as a new section “Parameters priority” with reformatted examples and a note about “Raw Query Parameters” having lesser priority than other form fields in commit 6b06a4c.
